content/advent-2016/gleam-distributed-map-reduce-for-golang.md

```diff
@@ -0,0 +1,356 @@
++++
+author = ["Chris Lu"]
+date = "2016-12-19T12:11:25-08:00"
+series = ["Advent 2016"]
+title = "Gleam: Distributed Map Reduce for Golang"
+
++++
+
+After developing [Glow](https://github.com/chrislusf/glow) last year,
+I came to realize the two limitations of Go for distributed data processing.
+
+First, generics are needed. Of course, we can use reflection. But it is 
+noticeably slower, to the point that I do not want to show the performance
+numbers. Second, dynamic remote code execution is also
+needed if we want to dynamically adjust the execution plan. We could pre-build
+all the execution DAGs first and choose one of them during run time. But it
+is very limiting.
+
+As everyone else here, I enjoyed the beauty of Go. How to make it work for
+big data?
+
+Then I found LuaJIT. It seems an ideal complement to Go. For many who do
+not know LuaJIT, its performance is on the same level as C, Java, Go. Being
+a scripting language, it perfectly resolves the two limitations above:
+generics and dynamic remote code execution. Lua was created as a small and
+embeddable language. Go should embrace it. (BTW, LuaJIT's FFI is really
+easy to call an external C library function, much simpler than CGO.)
+
+Additionally, Unix Pipe tools, e.g., "sort", "tr", "uniq", etc, 
+are also gems for data proccesing. But being single threaded, they lack 
+the power to tackle big data. Go should be able to help to scale them up,
+to distributedly run these tools.
+
+With [Gleam](https://github.com/chrislusf/gleam), we can combine pure Go, 
+LuaJIT, and Unix Pipes, all the 3 powerful weapons together.
+
+Let's understand Gleam architecutre of how it works. 
+Then I will cover 2 examples:
+
+1. Sort an 1GB file in both standalone and distributed modes, and distributed
+Unix sort mode, with performance comparison.
+2. Implement Joining of CSV files.
```

Comment:
`Joining` should be `joining`, or at least is there is a reason it's capitalised?